src/platform/qt/library/LibraryController.cpp

```diff
@@ -0,0 +1,185 @@
+/* Copyright (c) 2014-2017 waddlesplash
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+#include "LibraryController.h"
+
+#include "../GBAApp.h"
+#include "LibraryGrid.h"
+#include "LibraryTree.h"
+
+namespace QGBA {
+
+LibraryEntry::LibraryEntry(mLibraryEntry* entry)
+    : entry(entry)
+    , m_fullpath(QString("%1/%2").arg(entry->base, entry->filename))
+{
+}
+
+void AbstractGameList::addEntries(QList<LibraryEntryRef> items) {
+	for (LibraryEntryRef o : items) {
+		addEntry(o);
+	}
+}
+void AbstractGameList::removeEntries(QList<LibraryEntryRef> items) {
+	for (LibraryEntryRef o : items) {
+		addEntry(o);
+	}
+}
+
+LibraryLoaderThread::LibraryLoaderThread(QObject* parent)
+	: QThread(parent)
+{
+}
+
+void LibraryLoaderThread::run() {
+	mLibraryLoadDirectory(m_library, m_directory.toUtf8().constData());
+	m_directory = QString();
+}
+
+LibraryController::LibraryController(QWidget* parent, const QString& path, ConfigController* config)
+	: QStackedWidget(parent)
+	, m_config(config)
+{
+	mLibraryListingInit(&m_listing, 0);
+
+	if (!path.isNull()) {
+		m_library = mLibraryLoad(path.toUtf8().constData());
+	} else {
+		m_library = mLibraryCreateEmpty();
+	}
+
+	mLibraryAttachGameDB(m_library, GBAApp::app()->gameDB());
+
+	m_libraryTree = new LibraryTree(this);
+	addWidget(m_libraryTree);
+	connect(m_libraryTree, &LibraryTree::startGame, this, &LibraryController::startGame);
+
+	m_libraryGrid = new LibraryGrid(this);
+	addWidget(m_libraryGrid);
+	connect(m_libraryGrid, &LibraryGrid::startGame, this, &LibraryController::startGame);
+
+	connect(&m_loaderThread, &QThread::finished, this, &LibraryController::refresh, Qt::QueuedConnection);
+
+	refresh();
+
+	setViewStyle(LibraryStyle::STYLE_TREE);
+}
+
+LibraryController::~LibraryController() {
+	mLibraryListingDeinit(&m_listing);
+	if (m_library) {
```

Comment:
If the loader thread is active, make sure to wait on it finishing so you don't leak the m_library.